src/descriptors/element_clip_edge.js

```diff
@@ -0,0 +1,195 @@
+
+ElementClipEdge.top = factoryFn(TOP);
+ElementClipEdge.right = factoryFn(RIGHT);
+ElementClipEdge.bottom = factoryFn(BOTTOM);
+ElementClipEdge.left = factoryFn(LEFT);
+
+ElementClipEdge.prototype.value = function value() {
+	ensure.signature(arguments, []);
+
+	var clipPosition = this.getRawClipPosition();
+
+	if (!clipPosition) {
+		throw new ClipNotAppliedException(ElementClipEdge.prototype.value,
+			"clip " + this._position + " css style not applied to " + this._element);
+	}
+
+	var edge = clipPosition[this._position];
+	var scroll = this._element.frame.getRawScrollPosition();
+
+	if (this._position === RIGHT || this._position === LEFT) return Position.x(edge + scroll.x);
+	else return Position.y(edge + scroll.y);
+};
+
+ElementClipEdge.prototype.toString = function toString() {
+	ensure.signature(arguments, []);
+	return "clip " + this._position + " edge of " + this._element;
+};
+
+function factoryFn(position) {
+	return function factory(element) {
+		return new ElementClipEdge(element, position);
+	};
+}
+
+var UNSET_CLIP_STYLES = [ "", "auto", "unset", "initial" ];
+var CLIP_RECT_PATTERN = /rect[\s]*\([\s]*([^\s,]+)[\s,]+([^\s,]+)[\s,]+([^\s,]+)[\s,]+([^\s,]+)[\s]*\)/;
+
+ElementClipEdge.prototype.getRawClipPosition = function getRawClipPosition() {
+	ensure.signature(arguments, []);
+
+	var rect = this._element.getRawStyle("clip");
+
+	if (rect === "") {
+		// As a fallback for IE8 for when it can't fork over the original clip css style, try generating a clip rect
+		// using clip components that currentStyle may have.  If we see non-empty strings for all four components, we'll
+		// build out a clip rect expression here ...
+
+		var clipLeft = this._element.getRawStyle("clip-left");
+		var clipRight = this._element.getRawStyle("clip-right");
+		var clipBottom = this._element.getRawStyle("clip-bottom");
+		var clipTop = this._element.getRawStyle("clip-top");
+
+		if (clipLeft && clipRight && clipBottom && clipTop) {
+			rect = "rect(" + clipTop + " " + clipRight + " " + clipBottom + " " + clipLeft + ")";
+		}
+	}
+
+	for (var i = 0, ii = UNSET_CLIP_STYLES.length; i < ii; i++) {
+		if (UNSET_CLIP_STYLES[i] === rect) {
+			return null;
+		}
+	}
+
+	var matches = rect.match(CLIP_RECT_PATTERN);
+	if (!matches) {
+		ensure.unreachable("Unknown clip css style: " + rect);
+	}
+
+	// values in a clip's rect may be a css length or "auto" which means "clip over the edge's border"
+	var clipTopPx = this.computeClipTopPxHeight(matches[1]);
+	var clipRightPx = this.computeClipRightPxWidth(matches[2]);
+	var clipBottomPx = this.computeClipBottomPxHeight(matches[3]);
+	var clipLeftPx = this.computeClipLeftPxWidth(matches[4]);
+
+	var domElement = this._element.toDomElement();
+	var boundingRect = domElement.getBoundingClientRect();
+
+	var clipRect = {
+		left: boundingRect.left + clipLeftPx,
+		right: boundingRect.left + clipRightPx,
+
+		top: boundingRect.top + clipTopPx,
+		bottom: boundingRect.top + clipBottomPx
+	};
```

Comment:
Instead of calculating the element's bounding box yourself, rely on existing descriptors. This prevents duplication and allows you to take advantage of any existing normalization logic.

```javascript
var clipRect = {
  left: this.element.left.value().plus(clipLeftPx),
  right: this.element.right.value().plus(clipRightPx),
  ...
};